### ISSUE TITLE
Feature: native network load balancing

### ISSUE THREAD

**Issue by falzm** [balabit/syslog-ng#1210]
Hi

I'd like to know if it has been considered implementing a native network-related destination load balancing mecanism in syslog-ng? I'm running *loghosts* that receive logs from hundreds of servers: those loghosts are accessed through a DNS-based endpoint using Consul, however once a server is connected to a loghost it streams its logs only to this server until the connection dies or one of the sender/receiver syslog-ng instance is restarted. The problem is that the distribution of the logs is very unbalanced, and some loghosts are much more loaded than others because of the various logging throughputs across the source servers...

I've search for a bit and the closest I've found is a [mailing list thread](https://lists.balabit.hu/pipermail/syslog-ng/2011-September/017323.html) back in 2011 where the ad-hoc solution was to perform sort-of manual distribution by hashing/regexp'ing the `$SEQNUM` variable of log messages and choose one of the destinations based on the output of the hash/regexp – which is very hacky and I guess very inefficient.

It would be very useful to be able to specify several hosts in the network-based destinations – e.g. `syslog(ip("host1", "host2", "hostN") transport(tcp) ...)` – and have syslog-ng able to split the traffic at protocol level (i.e. on a message basis) among the configured hosts, say in a round-robin fashion, thus ensuring a much more fair logging traffic balancing.

**Comment by bazsi** [comment]
Agreed, this would be a great addition. It would probably have to work with
a $HOST based hashing though, as normally syslog messages are expected to
be logged in sequence, at least from the same host.

The "protocol implementation" is encapsulated into a LogWriter instance,
which the AFSocketDestDriver could use multiple of. A similar
implementation is in the affile module, where we do template based queueing
to multiple files.



-- 
Bazsi

**Comment by falzm** [comment]
@bazsi well, if it could be up to the user to choose the key on which the hashing would be performed that would be great, since hashing on `$HOST` would have the same unbalanced effect in my case... In my case it's OK for logs to be spread on multiple destinations.

**Comment by bazsi** [comment]
Hi,

ok, and you don't care about sequenced delivery within the same host? what
about "$HOST$PROGRAM$PID" would that resolve this for you? Although
PROGRAM/PID are not always filled.

-- 
Bazsi

**Comment by falzm** [comment]
In my use case I'm dealing with JSON-serialized messages, where a message is self-contained (i.e. no chance of being split on several consecutive messages). So as long as the user can specify the log element on which perform the hashing – for example `$SEQNUM` –  I'm good :)

**Comment by faxm0dem** [comment]
:+1: would be a killer feature. We're currently doing host-based hashing using puppet (talking about hackish ;))

**Comment by btorch** [comment]
👍

**Comment by presidento** [comment]
@falzm @faxm0dem @btorch what do you think about the builtin simple network load balancer implemented in #1706?
